### PR TITLE
Myriad Bluetooth and connectivity fixes

### PR DESCRIPTION
**Pull request by jwise**
There are three main thrusts of this pull request:

  * Implement a pairing database, and pairing UI.  The device goes into advertising mode when you enter the settings menu (which is created in this PR), and only accepts new pairing requests at that time.  Pairing is not quite implemented the way real Pebble devices do it, but this appears to be compatible with the Pebble app.
  * Shore up the PPoGATT (and, while we're at it, everything else Bluetooth) support, especially on nRF52.  This allows nRF52 to reliably talk to the Pebble app, whenever the Pebble app can be bothered to establish a connection, anyway.
  * Fix a few protocol-layer things, but not much.  The watch still can't, for instance, convince the Pebble app to send it an app list, and the watch doesn't have a way of recording what it was most recently connected to, so it can't mark itself as "faithful" or "unfaithful" to the phone (and avoid the full-database-wipe that is mandated in such an event).

### CODE REVIEW

File: Apps/System/settings.c [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+static Menu *_menu;
+static StatusBarLayer *status_bar;
+
+static void _reset_menu_items(void);
+
+
+static void exit_to_watchface(struct Menu *menu, void *context)
+{
+    /* Exit to watchface */
+    hw_bluetooth_advertising_visible(0);
```
Comment (ginge) [inline Apps/System/settings.c:95]: prefer if would could stub this through bluetooth.c.

File: Apps/System/settings.c [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+    window_stack_pop(true);
+}
+
+static struct MenuItems *_really_wipe_fs(const struct MenuItem *ctx) {
```
Comment (ginge) [inline Apps/System/settings.c:99]: #ifdef for platform support?

File: Apps/System/settings.c [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+
+static void _reset_menu_items(void)
+{
+    MenuItems *back = _menu->items->back;
+    MenuIndex *index = &_menu->items->back_index;
+
+    MenuItems *items = menu_items_create(4);
+    
+    menu_items_add(items, MenuItem("Discoverable", hw_bluetooth_name(), RESOURCE_ID_SPANNER, NULL));
+    menu_items_add(items, MenuItem("Format filesystem", "Time to die, sucker!", RESOURCE_ID_SPANNER, _wipe_fs));
```
Comment (ginge) [inline Apps/System/settings.c:130]: #ifdef for platform support here actually

File: hw/drivers/nrf52_bluetooth/nrf52_bluetooth_internal.h [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+ * Definitions for internal components of nRF52 Bluetooth driver, including PPoGATT.  Not for external use.
+ * RebbleOS
+ *
+ * Author: Joshua Wise <joshua@joshuawise.com>
+ */
+#pragma once
+
+#include "ble_db_discovery.h"
+
+#define PPOGATT_MTU 256
```
Comment (ginge) [inline hw/drivers/nrf52_bluetooth/nrf52_bluetooth_internal.h:11]: we should have a better mechanism for this. per platform MTU config

File: rcore/appmanager_app_api.c [hunk 5]
```diff
@@ -26,7 +30,11 @@ void appmanager_app_start(char *name)
 void appmanager_app_start_by_uuid(Uuid *uuid)
 {
     /* get app by uuid */
-    App * app = appmanager_get_app_by_uuid(uuid);   
+    App * app = appmanager_get_app_by_uuid(uuid);
+    if (!app) {
+        LOG_ERROR("appmanager was asked to start by uuid, but no such app existed");
+        return;
+    }
     assert(app);
```
Comment (ginge) [inline rcore/appmanager_app_api.c:38]: redundant?

File: rcore/bluetooth.c [hunk 6]
```diff
@@ -196,15 +212,13 @@ static void _bt_cmd_thread(void *pvParameters)
...
         if (pkt.packet_type == PACKET_TYPE_TX)
         {
-            BT_LOG("BT", APP_LOG_LEVEL_INFO, "TX %d byte", pkt.length);
-
             bt_device_request_tx(pkt.data, pkt.length);
             int rv = ulTaskNotifyTake(pdTRUE, pdMS_TO_TICKS(50));
             if(rv == 0)
                 BT_LOG("BT", APP_LOG_LEVEL_ERROR, "Timed out sending!");
 
             /* XXX ugh, should be using a better dynamic pool of memory not sys malloc */
```
Comment (ginge) [inline rcore/bluetooth.c:220]: dead comment is only half true

File: rcore/bluetooth.c [hunk 7]
```diff
@@ -257,7 +271,26 @@ inline uint8_t bluetooth_send(uint8_t *data, size_t len)
...
+    
+    /* Forge an rx ping to set the transport, and to transmit real data
+     * over the link (a pong).  */
+    static const uint8_t appversion_req[] = {
+        0x00
+    };
+    RebblePacket pkt = packet_create_with_data(WatchProtocol_AppVersion, appversion_req, sizeof(appversion_req));
+    packet_send(pkt);
+
+#if 0
```
Comment (ginge) [inline rcore/bluetooth.c:285]: weirdly this is still something that is needed on snowy

File: rcore/log.c [hunk 8]
```diff
@@ -32,97 +78,85 @@ void log_printf_to_ar(const char *layer, const char *module, uint8_t level, cons
...
- 
-    
 #define INT_LEN 6
 #define LEVEL_LEN 3
 #define LAYER_LEN 8
 #define MODULE_LEN 8
 #define FILENM_LEN 15
 #define LINENO_LEN 5
  
-    snprintf(buf, (INT_LEN / 2) + 1, "[%d]", interrupt_set);
```
Comment (ginge) [inline rcore/log.c:71]: :rocket: finally this shit is good

File: Apps/System/settings.c (hunk 2 above)
Comment (jwise) [reply Apps/System/settings.c:99]: This should be safe on Snowy too?  If it isn't, we oughta fix it.

File: Apps/System/settings.c (hunk 3 above)
Comment (jwise) [reply Apps/System/settings.c:130]: See above.  If this isn't safe on snowy, let's fix it.

File: Apps/System/settings.c (hunk 1 above)
Comment (jwise) [reply Apps/System/settings.c:95]: Ok, this and hw_bluetooth_name now go through bluetooth.c.  Reminds me that I want to take a hammer to the Bluetooth HAL / rcore interface at some point anyway.  But not now.  Other fish to fry for now.

File: rcore/bluetooth.c (hunk 7 above)
Comment (jwise) [reply rcore/bluetooth.c:285]: Ok, ifdef'ed for snowy.

File: rcore/bluetooth.c (hunk 6 above)
Comment (jwise) [reply rcore/bluetooth.c:220]: Removed!

File: hw/drivers/nrf52_bluetooth/nrf52_bluetooth_internal.h (hunk 4 above)
Comment (jwise) [reply hw/drivers/nrf52_bluetooth/nrf52_bluetooth_internal.h:11]: Yeah, I agree.  PPOGATT MTU is still busted in a lot of ways; MTU negotiation doesn't take place properly, for instance.  I don't really understand how it's supposed to work, and it's going to take some time before I do, I think.  In the mean time, this is a ... maximum ... MTU.

File: rcore/appmanager_app_api.c (hunk 5 above)
Comment (jwise) [reply rcore/appmanager_app_api.c:38]: Sure does!  FIxed.

**Pull request merged**